// music-engines/dist/src/soundcloud/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SoundcloudWrappers = exports.SoundCloud = void 0;
const utils_1 = require("./utils");
const wrappers_1 = require("./wrappers");
const soundcloud_scraper_1 = require("soundcloud-scraper");
class SoundCloud {
    constructor() {
        this.client = new soundcloud_scraper_1.Client();
        this.validator = utils_1.SoundcloudValidator;
        this.extractor = utils_1.SoundcloudExtractor;
    }
    use(input, type = 'all') {
        return new Promise((resolve, reject) => {
            const isValid = this.validator(input), extracted = this.extractor(input);
            if (isValid) {
                if (extracted.type == 'track') {
                    this.client.getSongInfo(input, {
                        fetchStreamURL: true
                    })
                        .then(data => {
                        resolve(new wrappers_1.SoundCloudTrack(data));
                    })
                        .catch(reject);
                }
                else if (extracted.type == 'playlist') {
                    this.client.getPlaylist(input)
                        .then(data => {
                        resolve(new wrappers_1.SoundCloudPlaylist(data));
                    })
                        .catch(reject);
                }
                else if (extracted.type == 'artist') {
                    this.client.getUser(input)
                        .then(data => {
                        resolve(new wrappers_1.SoundCloudArtist(data));
                    })
                        .catch(reject);
                }
            }
            else {
                this.client.search(input, type)
                    .then((data) => { var data_1, data_1_1; return __awaiter(this, void 0, void 0, function* () {
                    var e_1, _a;
                    let output = [];
                    try {
                        for (data_1 = __asyncValues(data); data_1_1 = yield data_1.next(), !data_1_1.done;) {
                            const item = data_1_1.value;
                            try {
                                if (item.type == 'track') {
                                    output.push(new wrappers_1.SoundCloudTrack(yield this.client.getSongInfo(item.url, {
                                        fetchStreamURL: true
                                    })));
                                }
                                else if (item.type == 'playlist') {
                                    output.push(new wrappers_1.SoundCloudPlaylist(yield this.client.getPlaylist(item.url)));
                                }
                                else if (item.type == 'artist') {
                                    output.push(new wrappers_1.SoundCloudArtist(yield this.client.getUser(item.url)));
                                }
                            }
                            catch (error) {
                                reject(error);
                            }
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (data_1_1 && !data_1_1.done && (_a = data_1.return)) yield _a.call(data_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                    console.log('For ended, Tracks: ', output);
                    resolve(output);
                }); })
                    .catch(reject);
            }
        });
    }
}
exports.SoundCloud = SoundCloud;
exports.SoundcloudWrappers = __importStar(require("./wrappers"));
