// music-engines/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Deezer = exports.SoundCloud = exports.Spotify = exports.YouTube = exports.wrappers = exports.typings = void 0;
const YTTypings = __importStar(require("./typings/youtube"));
const SPTypings = __importStar(require("./typings/spotify"));
const SCTypings = __importStar(require("./typings/soundcloud"));
const DZTypings = __importStar(require("./typings/deezer"));
const youtube_1 = require("./youtube");
Object.defineProperty(exports, "YouTube", { enumerable: true, get: function () { return youtube_1.YouTube; } });
const spotify_1 = require("./spotify");
Object.defineProperty(exports, "Spotify", { enumerable: true, get: function () { return spotify_1.Spotify; } });
const soundcloud_1 = require("./soundcloud");
Object.defineProperty(exports, "SoundCloud", { enumerable: true, get: function () { return soundcloud_1.SoundCloud; } });
const deezer_1 = require("./deezer");
Object.defineProperty(exports, "Deezer", { enumerable: true, get: function () { return deezer_1.Deezer; } });
const typings = {
    youtube: YTTypings,
    spotify: SPTypings,
    soundcloud: SCTypings,
    deezer: DZTypings
};
exports.typings = typings;
const wrappers = {
    youtube: youtube_1.YoutubeWrappers,
    spotify: spotify_1.SpotifyWrappers,
    soundcloud: soundcloud_1.SoundcloudWrappers,
    deezer: deezer_1.DeezerWrappers
};
exports.wrappers = wrappers;
