// music-engines/dist/src/spotify/wrappers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpotifyTrack = exports.SpotifyPlaylist = exports.SpotifyArtist = exports.SpotifyAlbum = void 0;
const spotify_url_info_1 = require("spotify-url-info");
const youtube_1 = require("../youtube");
const wrappers_1 = require("../youtube/wrappers");
const Base_1 = require("../Base");
class SpotifyTrack extends Base_1.Base {
    constructor(data) {
        super('spotify');
        this.raw = data;
        this.id = data.id;
        this.title = data.name;
        this.picture = data.album.images[0];
        this.url = data.external_urls.spotify;
        this.duration = this._DurationFormater(data.duration_ms, true);
    }
    _DurationFormater(data, isMs) {
        if (isMs) {
            var ms = data % 1000;
            data = (data - ms) / 1000;
        }
        var secs = data % 60;
        data = (data - secs) / 60;
        var mins = data % 60;
        var hrs = (data - mins) / 60;
        return {
            full: data,
            ms: (isMs) ? data : (data * 1000),
            sec: secs,
            min: mins,
            hour: hrs,
            format: (hrs) ? `${hrs}:${mins}:${secs}` : `${mins}:${secs}`
        };
    }
    stream(ytdlParams) {
        return new Promise((resolve, reject) => {
            let searchQuery = `${this.raw.artists[0].name} - ${this.raw.name}`;
            const YouTubeInstance = new youtube_1.YouTube();
            YouTubeInstance.use(searchQuery, {
                limit: 1,
                format: true,
                type: 'video'
            })
                .then(res => {
                const data = res[0];
                if (data instanceof wrappers_1.YouTubeTrack) {
                    resolve(data.stream(ytdlParams));
                }
            })
                .catch(reject);
        });
    }
}
exports.SpotifyTrack = SpotifyTrack;
class SpotifyAlbum extends Base_1.Base {
    constructor(data) {
        super('spotify');
        this.raw = data;
        this.url = data.external_urls.spotify;
        this.id = data.id;
        this.picture = data.images[0].url;
        this.title = data.name;
        this.releaseDate = data.release_date;
    }
    artists() {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var e_1, _a;
            let output = [];
            try {
                for (var _b = __asyncValues(this.raw.artists), _c; _c = yield _b.next(), !_c.done;) {
                    const art = _c.value;
                    output.push(new SpotifyArtist(art));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) yield _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            resolve(output);
        }));
    }
    tracks() {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            (0, spotify_url_info_1.getData)(this.raw.external_urls.spotify)
                .then((albumData) => __awaiter(this, void 0, void 0, function* () {
                var e_2, _a;
                let output = [];
                try {
                    for (var _b = __asyncValues(albumData.tracks.items), _c; _c = yield _b.next(), !_c.done;) {
                        const item = _c.value;
                        output.push(new SpotifyTrack(item));
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) yield _a.call(_b);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
                resolve(output);
            }))
                .catch(reject);
        }));
    }
}
exports.SpotifyAlbum = SpotifyAlbum;
class SpotifyArtist extends Base_1.Base {
    constructor(data) {
        super('spotify');
        this.raw = data;
        this.id = data.id;
        this.name = data.name;
        this.url = data.external_urls.spotify;
        this.socials = data.external_urls;
        this.picture = data.images[0];
    }
    tracks() {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var e_3, _a;
            let output = [];
            try {
                for (var _b = __asyncValues(this.raw.tracks), _c; _c = yield _b.next(), !_c.done;) {
                    const track = _c.value;
                    output.push(new SpotifyTrack(track));
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) yield _a.call(_b);
                }
                finally { if (e_3) throw e_3.error; }
            }
            resolve(output);
        }));
    }
}
exports.SpotifyArtist = SpotifyArtist;
class SpotifyPlaylist extends Base_1.Base {
    constructor(data) {
        super('spotify');
        this.raw = data;
        this.id = data.id;
        this.title = data.name;
        this.url = data.external_urls.spotify;
        this.description = data.description;
        this.picture = data.images[0].url;
        this.size = data.tracks.items.length;
        this.publisher = data.owner;
    }
    tracks() {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            var e_4, _a;
            let output = [];
            try {
                for (var _b = __asyncValues(this.raw.tracks.items), _c; _c = yield _b.next(), !_c.done;) {
                    const item = _c.value;
                    output.push(new SpotifyTrack(item));
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) yield _a.call(_b);
                }
                finally { if (e_4) throw e_4.error; }
            }
            resolve(output);
        }));
    }
}
exports.SpotifyPlaylist = SpotifyPlaylist;
