// music-engines/dist/src/youtube/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.YouTubeExtractor = exports.YouTubeValidator = void 0;
const ytdl_core_1 = require("ytdl-core");
const youtube_sr_1 = __importDefault(require("youtube-sr"));
function validator(input) {
    if ((0, ytdl_core_1.validateID)(input) || (0, ytdl_core_1.validateURL)(input))
        return true;
    else
        return false;
}
exports.YouTubeValidator = validator;
function extractor(url) {
    let res = {
        url: ((0, ytdl_core_1.validateURL)(url)) ? url : `https://www.youtube.com/watch?v=${url}`,
        id: ((0, ytdl_core_1.validateURL)(url)) ? (0, ytdl_core_1.getVideoID)(url) : url,
        type: 'video'
    };
    if (youtube_sr_1.default.validate(res.url, 'VIDEO') || youtube_sr_1.default.validate(res.url, 'VIDEO_ID'))
        res.type = 'video';
    else if (youtube_sr_1.default.validate(res.url, 'PLAYLIST') || youtube_sr_1.default.validate(res.url, 'PLAYLIST_ID'))
        res.type = 'playlist';
    return res;
}
exports.YouTubeExtractor = extractor;
